File: ios/Sources/MoQKit/Core/Session.swift
import Foundation
import MoQKitFFI

// MARK: - SessionError (codec/format related)

/// High-level errors surfaced by relay connection, subscription, and playback setup APIs.
public enum SessionError: Error, Sendable, Equatable {
    /// The track uses a codec that MoQKit does not support.
    case unsupportedCodec(String)
    /// A video codec requires an out-of-band parameter set (SPS/PPS/VPS) but none was provided.
    case missingCodecDescription
    /// `CMFormatDescription` creation failed with the given OS status code.
    case formatDescriptionFailed(OSStatus)
    /// `CMSampleBuffer` creation failed with the given OS status code.
    case sampleBufferFailed(OSStatus)
    /// ``Session/connect()`` was called on a session that is already connecting or connected.
    case alreadyConnected
    /// ``Session/connect()`` was called after ``Session/close()`` — create a new session instead.
    case alreadyClosed
    /// ``Session/subscribe(prefix:)`` was called for a prefix that is already active.
    case alreadySubscribed
    /// No tracks were found in the broadcast catalog.
    case noTracksAvailable
    /// No broadcast was found at the given path.
    case noBroadcastAvailable
    /// ``Player`` was initialised with both video and audio disabled.
    case noTracksSelected
    /// A configuration invariant was violated (details in the associated string).
    case invalidConfiguration(String)
    /// The audio decoder failed to initialise or decode a frame.
    case audioDecoderFailed(String)
    /// The underlying QUIC/WebTransport connection failed. The associated string contains
    /// the transport-level error description.
    case connectionFailed(String)
}

// MARK: - State

/// The lifecycle state of a ``Session``.
public enum SessionState: Sendable, Equatable {
    /// Initial state. ``Session/connect()`` has not been called yet.
    case idle
    /// QUIC handshake is in progress.
    case connecting
    /// Transport is ready. The session may now publish and create broadcast subscriptions.
    case connected
    /// An irrecoverable error occurred. The session cannot be reused — create a new one.
    case error(SessionError)
    /// The session was closed via ``Session/close()``. No further events will be emitted.
    case closed
}

// MARK: - Session

/// Manages one MoQ relay connection and the broadcasts attached to it.
///
/// `Session` is the main entry point for MoQKit. Create one with a relay URL, call
/// ``connect()``, then use the same session to subscribe to broadcasts, publish tracks,
/// or both. After ``close()`` finishes, create a new session instead of reusing the old
/// instance.
///
/// To discover live streams, call ``subscribe(prefix:)`` after connecting and observe the
/// returned subscription:
///
/// ```swift
/// let session = Session(url: "https://relay.example.com/moq")
/// try await session.connect()
/// let subscription = try session.subscribe()
///
/// for await broadcast in subscription.broadcasts {
///     for await catalog in broadcast.catalogs() {
///         let videoTrack = catalog.playableVideoTracks.first?.name
///         let audioTrack = catalog.playableAudioTracks.first?.name
///         guard videoTrack != nil || audioTrack != nil else { continue }
///
///         let player = try await MainActor.run {
///             try Player(
///                 catalog: catalog,
///                 videoTrackName: videoTrack,
///                 audioTrackName: audioTrack
///             )
///         }
///         try await player.play()
///     }
/// }
/// ```
///
/// To publish without consuming, simply omit the ``subscribe(prefix:)`` call:
///
/// ```swift
/// let session = Session(url: "https://relay.example.com/moq")
/// try await session.connect()
/// try session.publish(path: "live/my-stream", publisher: publisher)
/// try await publisher.start()
/// ```
public actor Session {
    /// Emits the current ``SessionState`` and every subsequent state change.
    ///
    /// The stream always yields `.idle` as its first element. It completes when the
    /// session reaches `.closed`.
    public nonisolated let state: AsyncStream<SessionState>

    private let url: String

    private let stateContinuation: AsyncStream<SessionState>.Continuation
    private var currentState: SessionState = .idle

    // Pipeline objects
    private var client: MoqClient?
    private var consumeOrigin: MoqOriginProducer?
    private var publishOrigin: MoqOriginProducer?
    private var session: MoqSession?

    // Per-path publish state
    private var activePublishers: [String: Publisher] = [:]
    private var activeSubscriptions: [String: BroadcastSubscription] = [:]

    // Background tasks
    private var sessionMonitorTask: Task<Void, Never>?

    /// Creates a new session.
    ///
    /// - Parameter url: The WebTransport URL of the MoQ relay (e.g. `"https://relay.example.com/moq"`).
    public init(url: String) {
//        do {
//            try moqLogLevel(level: "TRACE")
//        } catch {}
        self.url = url

        var stateCont: AsyncStream<SessionState>.Continuation!
        self.state = AsyncStream { stateCont = $0 }
        self.stateContinuation = stateCont

        stateContinuation.yield(.idle)
    }

    /// Establishes the QUIC connection to the relay.
///
    /// Transitions the session through `.connecting` to `.connected`. After this succeeds,
    /// you can start watching broadcasts with ``subscribe(prefix:)`` and register
    /// publishers with ``publish(path:publisher:)``.
    ///
    /// - Throws: ``SessionError/alreadyConnected`` if called while connecting or connected.
    /// - Throws: ``SessionError/alreadyClosed`` if the session has already been closed.
    /// - Throws: ``SessionError/connectionFailed(_:)`` if the transport handshake fails.
    public func connect() async throws {
        guard currentState == .idle else {
            if currentState == .closed { throw SessionError.alreadyClosed }
            throw SessionError.alreadyConnected
        }

        KitLogger.session.debug("Connecting to \(self.url)")
        transition(to: .connecting)

        do {
            // 1. Create separate origins for consuming and publishing
            let consumeOrigin = MoqOriginProducer()
            self.consumeOrigin = consumeOrigin

            let publishOrigin = MoqOriginProducer()
            self.publishOrigin = publishOrigin

            let client = MoqClient()
            client.setTlsDisableVerify(disable: true)
            client.setConsume(origin: consumeOrigin)
            client.setPublish(origin: publishOrigin)
            self.client = client

            // 2. Connect session
            let session = try await client.connect(url: url)
            self.session = session

            // 3. Connection established
            transition(to: .connected)

            // 4. Monitor session lifetime
            sessionMonitorTask = Task { [weak self] in
                do {
                    try await session.closed()
                } catch {
                    guard let self else { return }
                    await self.handleSessionEnded(error: error)
                    return
                }
                guard let self else { return }
                await self.handleSessionEnded(error: nil)
            }

        } catch let error as MoqError {
            let sessionError = SessionError.connectionFailed(error.moqKitMessage)
            KitLogger.session.error("Connection failed: \(sessionError)")
            transition(to: .error(sessionError))
            tearDown()
            throw sessionError
        } catch let error as SessionError {
            KitLogger.session.error("Connection failed: \(error)")
            transition(to: .error(error))
            tearDown()
            throw error
        } catch {
            let sessionError = SessionError.connectionFailed(error.localizedDescription)
            KitLogger.session.error("Connection failed: \(sessionError)")
            transition(to: .error(sessionError))
            tearDown()
            throw error
        }
    }

    /// Starts watching for broadcast announcements on the relay.
///
    /// Call this after ``connect()`` to receive broadcasts whose path starts with `prefix`.
    /// Keep the returned subscription alive for as long as you want updates, then call
    /// ``BroadcastSubscription/cancel()`` when that part of your app is done.
    ///
    /// - Parameter prefix: Only broadcasts whose path starts with this string will be surfaced.
    ///   Pass `""` (the default) to receive all broadcasts.
    /// - Throws: ``SessionError/invalidConfiguration(_:)`` if the session is not connected.
    /// - Throws: ``SessionError/alreadySubscribed`` if the exact prefix is already active.
    public func subscribe(prefix: String = "") throws -> BroadcastSubscription {
        guard currentState == .connected else {
            throw SessionError.invalidConfiguration(
                "Session must be connected before subscribing")
        }
        guard let consumeOrigin else {
            throw SessionError.invalidConfiguration("Consume origin not available")
        }
        if let existingSubscription = activeSubscriptions[prefix] {
            if existingSubscription.isFinished {
                activeSubscriptions.removeValue(forKey: prefix)
            } else {
                throw SessionError.alreadySubscribed
            }
        }

        KitLogger.session.debug("Subscribing to announcements with prefix: \(prefix)")

        let consumer = consumeOrigin.consume()
        let announced = try consumer.announced(prefix: prefix)
        let subscription = BroadcastSubscription(
            prefix: prefix,
            session: self,
            announced: announced
        )
        activeSubscriptions[prefix] = subscription
        return subscription
    }

    /// Registers a broadcast publisher at the given relay path.
///
    /// This makes the publisher available on the relay, but it does not start your capture
    /// sources or encoders. Call ``Publisher/start()`` after this to begin sending frames.
    ///
    /// - Parameters:
    ///   - path: The broadcast path on the relay (e.g. `"live/my-stream"`).
    ///   - publisher: A configured ``Publisher`` with at least one track added.
    /// - Throws: ``SessionError/invalidConfiguration(_:)`` if the session is not connected.
    public func publish(path: String, publisher: Publisher) throws {
        guard currentState == .connected else {
            throw SessionError.invalidConfiguration(
                "Session must be connected before publishing")
        }
        guard let publishOrigin else {
            throw SessionError.invalidConfiguration("Publish origin not available")
        }
        KitLogger.publish.debug("Publishing broadcast at path: \(path)")
        try publishOrigin.publish(path: path, broadcast: publisher.broadcast)
        activePublishers[path] = publisher
    }

    /// Stops publishing at the given path.
///
    /// If a publisher is registered for that path, this calls ``Publisher/stop()`` and
    /// removes it from the session. Other subscriptions and publish paths continue running.
    public func unpublish(path: String) {
        guard let publisher = activePublishers.removeValue(forKey: path) else { return }
        KitLogger.publish.debug("Unpublishing broadcast at path: \(path)")
        publisher.stop()
    }

    /// Closes the relay connection and releases all resources.
    ///
    /// Transitions the session to `.closed` and completes ``state``.
    /// Safe to call multiple times — subsequent calls are no-ops.
    public func close() async {
        guard currentState != .closed else { return }
        KitLogger.session.debug("Closing session")
        tearDown()
        transition(to: .closed)
        stateContinuation.finish()
    }

    deinit {
        sessionMonitorTask?.cancel()

        let subscriptions = Array(activeSubscriptions.values)
        activeSubscriptions.removeAll()
        for subscription in subscriptions {
            subscription.cancel()
        }

        for (_, publisher) in activePublishers {
            publisher.stop()
        }
        activePublishers.removeAll()

        session?.cancel(code: 0)
        client?.cancel()
        stateContinuation.finish()
    }

    // MARK: - Private

    private func transition(to newState: SessionState) {
        KitLogger.session.debug(
            "State: \(String(describing: self.currentState)) → \(String(describing: newState))")
        currentState = newState
        stateContinuation.yield(newState)
    }

    func removeSubscription(prefix: String, matching subscription: BroadcastSubscription) {
        guard activeSubscriptions[prefix] === subscription else { return }
        activeSubscriptions.removeValue(forKey: prefix)
    }

    private func handleSessionEnded(error: Error?) async {
        if let error {
            KitLogger.session.warning("Session ended with error: \(error)")
            transition(to: .error(.connectionFailed("Session ended: \(error.localizedDescription)")))
            await close()
            return
        }

        if currentState == .connected {
            KitLogger.session.warning("Session ended unexpectedly")
            transition(to: .error(.connectionFailed("Session ended unexpectedly")))
            await close()
        }
    }

    private func tearDown() {
        KitLogger.session.debug("Tearing down session")

        sessionMonitorTask?.cancel()
        sessionMonitorTask = nil

        let subscriptions = Array(activeSubscriptions.values)
        activeSubscriptions.removeAll()
        for subscription in subscriptions {
            subscription.cancel()
        }

        for (_, publisher) in activePublishers { publisher.stop() }
        activePublishers.removeAll()

        session?.cancel(code: 0)
        session = nil

        client?.cancel()
        client = nil

        consumeOrigin = nil
        publishOrigin = nil
    }
}

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/Session.kt
package com.swmansion.moqkit

import android.util.Log
import com.swmansion.moqkit.publish.Publisher
import com.swmansion.moqkit.subscribe.BroadcastSubscription
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import uniffi.moq.MoqClient
import uniffi.moq.MoqOriginProducer
import uniffi.moq.MoqSession as UniMoqSession

/**
 * A connection to one MoQ relay.
 *
 * Create one session for a relay URL and use it as the starting point for discovering
 * broadcasts, publishing media, and sending or receiving raw data tracks. A session owns
 * the native connection and the work created through it, so closing the session also
 * closes active subscriptions and stops active publishers.
 *
 * ### Lifecycle
 * 1. Create a session with the relay [url].
 * 2. Call [connect] to establish the connection.
 * 3. Call [subscribe] to discover live broadcasts, and/or [publish] to announce media.
 * 4. Call [close] to tear down the connection and free all resources.
 *
 * @param url Relay URL, for example `"https://relay.example.com:4443/anon"`.
 * @param parentScope Coroutine scope whose lifetime bounds background session work. In apps,
 *   pass a lifecycle-owned scope such as `lifecycleScope` or `viewModelScope`.
 */
class Session(
    private val url: String,
    parentScope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob()),
) {
    companion object {
        private const val TAG = "Session"
    }

    private val scope = CoroutineScope(parentScope.coroutineContext + SupervisorJob())

    /** Connection state machine for this session. */
    sealed class State {
        /** Session has not been started yet. */
        object Idle : State()
        /** QUIC handshake is in progress. */
        object Connecting : State()
        /** Handshake complete; the session may now publish and subscribe. */
        object Connected : State()
        /** The session ended due to a transport or protocol error.
         * @property message Human-readable error description. */
        data class Error(val message: String) : State()
        /** [Session.close] was called and all resources have been released. */
        object Closed : State()
    }

    private val _state = MutableStateFlow<State>(State.Idle)

    /**
     * Current connection state. Starts at [State.Idle] and progresses through
     * [State.Connecting] → [State.Connected] → [State.Closed] (or [State.Error]).
     */
    val state: StateFlow<State> = _state.asStateFlow()

    private var session: UniMoqSession? = null
    private var client: MoqClient? = null
    private var consumeOrigin: MoqOriginProducer? = null
    private var publishOrigin: MoqOriginProducer? = null
    private var monitorJob: Job? = null

    private val activeSubscriptions = mutableMapOf<String, BroadcastSubscription>()
    private val activePublishers = mutableMapOf<String, Publisher>()

    /**
     * Opens the QUIC connection.
     *
     * Suspends until the handshake is complete. Once this function returns, [state] is
     * [State.Connected] and the session may publish or create broadcast subscriptions.
     *
     * A session can be connected only once. Create a new [Session] after [close].
     *
     * @throws IllegalStateException if called on a session that has already been started.
     * @throws Exception if the connection attempt fails (state becomes [State.Error]).
     */
    suspend fun connect() {
        check(_state.value == State.Idle) { "Session already started" }
        _state.value = State.Connecting
        Log.d(TAG, "Connecting to $url")
        try {
            val newConsumeOrigin = MoqOriginProducer()
            consumeOrigin = newConsumeOrigin
            Log.d(TAG, "Consume origin created")

            val newPublishOrigin = MoqOriginProducer()
            publishOrigin = newPublishOrigin
            Log.d(TAG, "Publish origin created")

            val newClient = MoqClient()
            newClient.setTlsDisableVerify(true)
            client = newClient
            newClient.setConsume(newConsumeOrigin)
            newClient.setPublish(newPublishOrigin)

            val newSession = newClient.connect(url)
            session = newSession
            _state.value = State.Connected
            Log.d(TAG, "Connected successfully")

            monitorJob = scope.launch {
                try {
                    newSession.closed()
                } catch (e: Exception) {
                    Log.w(TAG, "Session closed with error: $e")
                    _state.compareAndSet(State.Connected, State.Error("Session ended: $e"))
                    close()
                    return@launch
                }

                if (_state.value == State.Connected) {
                    Log.w(TAG, "Session ended unexpectedly")
                    _state.value = State.Error("Session ended unexpectedly")
                    close()
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Connection failed: ${e.message}", e)
            _state.value = State.Error(e.message ?: "Connection failed")
            tearDown()
            throw e
        }
    }

    /**
     * Starts watching for broadcast announcements under the supplied prefix.
     *
     * Collect [BroadcastSubscription.broadcasts] to receive each matching broadcast as it is
     * announced by the relay. Each prefix may have only one active subscription on a session.
     *
     * @param prefix Only broadcasts whose path starts with this string will be surfaced.
     *   Pass `""` (the default) to receive all broadcasts.
     * @throws IllegalStateException if the session is not connected or the exact prefix
     *   already has an active subscription.
     */
    fun subscribe(prefix: String = ""): BroadcastSubscription {
        check(_state.value == State.Connected) { "Session must be connected before subscribing" }

        return synchronized(activeSubscriptions) {
            check(_state.value == State.Connected) {
                "Session must be connected before subscribing"
            }

            val existing = activeSubscriptions[prefix]
            check(existing == null || existing.isClosed) {
                "Already subscribed to prefix '$prefix'"
            }

            val sessionConsumeOrigin = consumeOrigin
                ?: error("Consume origin not available")
            val originConsumer = sessionConsumeOrigin.consume()
            try {
                val announced = originConsumer.announced(prefix)
                var subscriptionRef: BroadcastSubscription? = null
                val subscription = BroadcastSubscription(
                    prefix = prefix,
                    originConsumer = originConsumer,
                    announced = announced,
                    onClosed = {
                        synchronized(activeSubscriptions) {
                            if (activeSubscriptions[prefix] === subscriptionRef) {
                                activeSubscriptions.remove(prefix)
                            }
                        }
                    },
                )
                subscriptionRef = subscription
                activeSubscriptions[prefix] = subscription
                subscription
            } catch (t: Throwable) {
                try {
                    originConsumer.close()
                } catch (closeError: Exception) {
                    Log.w(TAG, "Failed to close origin consumer for prefix '$prefix'", closeError)
                }
                throw t
            }
        }
    }

    /**
     * Announces a publisher at the given broadcast path.
     *
     * Configure the [Publisher] first, call this method to register the broadcast with the
     * relay, then call [Publisher.start] to begin sending media or data.
     *
     * @param path Broadcast path on the relay (e.g. `"live/my-stream"`).
     * @param publisher A configured [Publisher] with at least one track added.
     * @throws IllegalStateException if the session is not connected or already publishes at
     *   the same [path].
     */
    fun publish(path: String, publisher: Publisher) {
        check(_state.value == State.Connected) { "Session must be connected before publishing" }
        check(!activePublishers.containsKey(path)) {
            "Already publishing at '$path'. Call unpublish() first."
        }
        val origin = publishOrigin ?: error("Publish origin not available")
        Log.d(TAG, "Publishing broadcast at '$path'")
        origin.publish(path, publisher.broadcast)
        activePublishers[path] = publisher
    }

    /**
     * Stops publishing at the given path.
     *
     * If [path] is active, this calls [Publisher.stop] for the associated publisher. If the
     * path is not active, this method does nothing.
     */
    fun unpublish(path: String) {
        val publisher = activePublishers.remove(path) ?: return
        Log.d(TAG, "Unpublishing broadcast at '$path'")
        publisher.stop()
    }

    /**
     * Closes the session and releases all resources.
     *
     * Safe to call from any thread. No-op if the session is already [State.Closed].
     * After this returns, [state] is [State.Closed] and background work owned by this
     * session is cancelled.
     */
    fun close() {
        val wasConnected = _state.compareAndSet(State.Connected, State.Closed)
        val wasConnecting = if (!wasConnected) {
            _state.compareAndSet(State.Connecting, State.Closed)
        } else {
            false
        }
        val wasError = if (!wasConnected && !wasConnecting) {
            val current = _state.value
            if (current is State.Error) {
                _state.compareAndSet(current, State.Closed)
            } else {
                false
            }
        } else {
            false
        }

        if (!wasConnected && !wasConnecting && !wasError) {
            Log.d(TAG, "close() called but already in state ${_state.value}")
            return
        }

        Log.d(
            TAG,
            "Closing session (was: connected=$wasConnected connecting=$wasConnecting error=$wasError)",
        )
        tearDown()
    }

    private fun tearDown() {
        val subscriptions = synchronized(activeSubscriptions) {
            activeSubscriptions.values.toList().also { activeSubscriptions.clear() }
        }
        Log.d(TAG, "Tearing down: ${subscriptions.size} active subscriptions")

        subscriptions.forEach { subscription ->
            try {
                subscription.close()
            } catch (e: Exception) {
                Log.w(TAG, "Failed to close broadcast subscription '${subscription.prefix}'", e)
            }
        }

        activePublishers.values.forEach { publisher ->
            try {
                publisher.stop()
            } catch (e: Exception) {
                Log.w(TAG, "Failed to stop publisher during teardown", e)
            }
        }
        activePublishers.clear()

        monitorJob?.cancel()
        monitorJob = null

        session?.cancel(0u)
        try {
            session?.close()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to close session", e)
        }
        session = null

        client?.cancel()
        try {
            client?.close()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to close client", e)
        }
        client = null

        try {
            publishOrigin?.close()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to close publish origin", e)
        }
        publishOrigin = null

        try {
            consumeOrigin?.close()
        } catch (e: Exception) {
            Log.w(TAG, "Failed to close consume origin", e)
        }
        consumeOrigin = null

        scope.cancel()
        Log.d(TAG, "Teardown complete")
    }
}
